From: David Woodhouse <dwmw@amazon.co.uk>
Date: Thu, 11 Jan 2018 21:46:28 +0000
Subject: xen/x86/retpoline/entry: Convert entry assembler indirect jumps -- ia32
Patch-mainline: Never, SUSE-Xen specific
References: bsc#1068032 CVE-2017-5715

In this patch:
patches.arch/08-x86-retpoline-entry-Convert-entry-assembler-indirect.patch
we converted 32bit and 64bit entries to retpolines. But we completely
omitted the ia32 emulation on 64bit. So fix this in this followup patch.

Note: this code was converted to C in 4.4.

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
Automatically created from "patches.arch/08-x86-retpoline-entry-Convert-entry-assembler-indirect-ia32.patch" by xen-port-patches.py

--- a/arch/x86/ia32/ia32entry-xen.S
+++ b/arch/x86/ia32/ia32entry-xen.S
@@ -15,6 +15,7 @@
 #include <asm/irqflags.h>
 #include <linux/linkage.h>
 #include <asm/spec_ctrl.h>
+#include <asm/nospec-branch.h>
 
 /* Avoid __ASSEMBLER__'ifying <linux/audit.h> just for this.  */
 #include <linux/elf-em.h>
@@ -328,7 +329,12 @@ ia32_do_call:
 	andq %r10,%rax
 	IA32_ARG_FIXUP
 .Lia32_dispatch:
+#ifdef CONFIG_RETPOLINE
+	mov ia32_sys_call_table(,%rax,8), %rax
+	call __x86_indirect_thunk_rax
+#else
 	call *ia32_sys_call_table(,%rax,8) # xxx: rip relative
+#endif
 ia32_sysret:
 	movq %rax,RAX-ARGOFFSET(%rsp)
 	CLEAR_RREGS -ARGOFFSET
@@ -398,7 +404,7 @@ ENTRY(ia32_ptregs_common)
 	CFI_REL_OFFSET	rsp,RSP-ARGOFFSET
 /*	CFI_REL_OFFSET	ss,SS-ARGOFFSET*/
 	SAVE_REST
-	call *%rax
+	CALL_NOSPEC %rax
 	RESTORE_REST
 	jmp  ia32_sysret	/* misbalances the return cache */
 	CFI_ENDPROC
